jmeter/src/main/java/org/eclipse/hono/jmeter/client/HonoReceiver.java

```diff
@@ -96,43 +103,91 @@ private void createConsumer() throws InterruptedException {
-                if (creationHandler.failed()) {
-                    LOGGER.error("HonoClient.createEventConsumer() failed", creationHandler.cause());
-                }
-                receiverLatch.countDown();
-            });
+            amqpNetworkClient.createEventConsumer(sampler.getTenant(), Integer.parseInt(sampler.getPrefetch()),
+                    this::messageReceived, creationHandler -> {
+                        if (creationHandler.failed()) {
+                            LOGGER.error("HonoClient.createEventConsumer() failed", creationHandler.cause());
+                        }
+                        receiverLatch.countDown();
+                    });
         }
         receiverLatch.await();
     }
 
-    public void sample(final SampleResult result, final boolean isUseSenderTime) {
+    public void sample(final SampleResult result) {
         synchronized (lock) {
-            long elapsed = System.currentTimeMillis() - sampleStart;
+            long elapsed = 0;
             result.setResponseCodeOK();
             result.setSuccessful(true);
             result.setSampleCount(messageCount);
-            result.setResponseMessage(MessageFormat.format("count: {0}, bytes received: {1}, period: {2}", messageCount, messageSize, elapsed));
             result.setBytes(messageSize);
-            if (messageCount > 0) {
+            result.setResponseMessage(
+                    MessageFormat.format("count: {0}, bytes received: {1}, period: {2}", messageCount,
+                            messageSize, elapsed));
+            if (sampler.isUseSenderTime()) {
+                if (messageCount > 1) {
+                    LOGGER.warn(
+                            "More than one message per receiver was not expected. It could result in wrong sampling time");
+                } else if (messageCount == 1) {
+                    elapsed = sampleEnd - sampleStart;
+                    result.setStampAndTime(sampleStart, elapsed);
+                } else {
+                    noMessagesReceived(result);
+                }
+            } else if (messageCount > 0 && sampleStart != 0) {
+                elapsed = System.currentTimeMillis() - sampleStart;
                 result.setStampAndTime(sampleStart, elapsed);
+                result.setIdleTime(elapsed);
             } else {
-                result.setStampAndTime(sampleStart, 0);
+                noMessagesReceived(result);
             }
-            LOGGER.info("{}: received batch of {} messages in {} milliseconds", sampler.getThreadName(), messageCount, elapsed);
+            LOGGER.info("{}: received batch of {} messages in {} milliseconds", sampler.getThreadName(), messageCount,
+                    elapsed);
             messageSize = 0;
             messageCount = 0;
-            sampleStart = System.currentTimeMillis();
+            sampleStart = 0;
+        }
+    }
```

Comment:
Could you explain the changed behavior? I do not get it. Why are no more then 1 messages expected from one sample to an other? 

In the original version the sampleStart was set with the first received massage (or the first after a sample). Then it is set to this time or the millis in the application property to have the sender time. With the next sample start the elapsed time is calculated for all samples that have been arrived.

Reply:
Assume, the jmeter scenario contains a "delay" between the message sent and message received, then in your case, the sampling time will also include this delay leading to wrong report. I believe, when we use the time stamp from the message, it has to be per message or we sample for all messages received without reading the time stamp from the message. So it is better to calculate the sampling time for each message as each message contains the time stamp.

Reply:
Why should the delay from the JMeter scenario should influence this? The sender sends a message (or 8 or 68..) in one sample. If after this a delay is in place, this has no influence to the send messages. The receiver receives the messages all the time, independent of his sampling. It does only calculate the results from the received messages (between two samples) at the sample time.

Reply:
My concern is as follows. For ease of discussion, I just use time stamp in string format HH:MM below.
```
Sender Message-1: { timeStamp : "09:45" }
Message-1 received : "9.46"
Sample invoked: sampling time is "9:46 - 9:45" = 1 seconds. 
```
Everything is good so far. But what is the sampling time for the below scenario

```
Sender Message-1: { timeStamp : "09:45" }
Message-1 received : "9.46"
Sender Message-2: { timeStamp : "09:47" }
Some other message received and not Message-2: "09:48" (no time stamp in payload)
Sample invoked: sampling time is "09:48 - 09:45" = 3 seconds
```

If Message-2 is correctly received, then we have no sampling for the message-1

```
Sender Message-1: { timeStamp : "09:45" }
Message-1 received : "9.46"
Sender Message-2: { timeStamp : "09:47" }
Message-2 received: "09:48" 
Sample invoked: sampling time is "09:48 - 09:47" = 1 second
```

Actually, I am not sure whether it is possible to sample more messages in one receiver.

Reply:
Ah okay - so you have an other scenario in mind, then I had for the JMeter Plugin. I wanted to do load-tests, with an exclusive address for this tests and constant full load.

To fulfill your scenario I would imagine to calculate the delivery time for the received message (message put in property/body by sender minus current time) and add it to a sampleDeliveryTime. Than at each sample this could be divided by count of received messages - so you get a mean time for the n messages. 

If you really want to do this for each message to get a min/max/mean you need to remember this and it would be more complex. But I think this are different scenarios. Maybe it would be better do separate this different topics in different PRs - then it would be much clearer?
1. PR - allow sender time in body
2. PR - correct delivery time calculation without full load (and start time set by sender)
3. PR - see below

WDYT?

And the construct with the messageCount==1 is to get the time for exactly one message? Maybe a third scenario? Would it not be easier to modify the sender for this, that the actual MAX_MESSAGES_PER_BATCH_SEND could be configured by user and also set to 1. Then the JMeter should not run a fixed time but only 1 sample = 1 message?

Reply:
I like the idea os sampleDeliveryTime of all messages divided by count of received messages. One could then also this JMeter plugin with time stamp in a load test. Sounds good.

And your 3 PR approach also sounds feasible.

Regarding your 3rd PR, you want to stop listening for messages once the count has reached. Did I understanding correctly?

Reply:
I meant that at the moment we can not send only one message, since there is a batch constant of 300, which will be sendet with each sample call of the sender - this was introduced to reduce the overhead of JMeter. I think this constant should be configurable. And if you configure it to the value 1 and build a JMeter plan, that sends 1 message and receives 1 message (not based on time) - that is would be easy so get the time for this message and also the plan could give the min/max/mean for n calls if configured for more than 1 in a loop.

Reply:
ok, Thanks for clarifying.

Reply:
You agreed to split this to separate PRs, are you? Then it might be better to close this one?

Reply:
@balsmn - how do we want to proceed with this PR? Do you want to create separate PRs?

Reply:
@pellmann I am ok with PR 2 and PR 3. 

I am only confused how I am to proceed for PR 1 (without full load). What would be the sampling time. Is it always last message received time - time in property/body ?

Reply:
As first try, I have pushed the changes you requested for PR 1 at the following branch. Please let me know if it is what you were expecting. If yes, then please close this PR and I will create a new PR for the following branch

https://github.com/bsinno/hono/tree/PR/jmeter_use_sender_timer_1

Reply:
@balsmn - the branch looks good to me!

Reply:
@pellmann I am now working on the 3rd PR you have suggested and have one question to it.

>> And the construct with the messageCount==1 is to get the time for exactly one message? Maybe a third scenario? Would it not be easier to modify the sender for this, that the actual MAX_MESSAGES_PER_BATCH_SEND could be configured by user and also set to 1. Then the JMeter should not run a fixed time but only 1 sample = 1 message?

Now that we don't have messageCount == 1 and we can't acutally control when the sample() method will be called, what would be the best way for the Hono receiver to react when it has reached MAX_MESSAGES_PER_BATCH_SEND and has received more messages? 

Option 1, is to ignore and don't include them in any measurements or 
Option 2 is to throw an exception (which in my opinion is also not correct as the Hono receiver has received all the messages it was expected to receive).

Are you ok with Option 1?